### DISCUSSION TITLE
Best practice for creating new apps

### DISCUSSION THREAD

**Discussion by phanpatrik** [mastodon/mastodon#20833] [Q&A]
I am currently creating an alternative web client for Mastodon, where users can log in using their server specific trade.

However, I am wondering what the best practice is when creating an app by using API (`/api/v1/apps`)

Should an app be registered exclusively **per instance**, allowing multiple users to authenticate on that server, or should it be generated **per user per instance** at best? So it could happen that the "same" app is registered several times on one server but with individual callback URLs etc.

Would love to get your take on this, so that I can design the auth process for a general web client.

Thanks

**Comment by kamcma** [comment]
I have also been searching for more information on this point.

It _seems_ like the answer is register an app per user per instance. _If_ Mastodon were to, say, de-duplicate app registrations behind the scenes and return the same client ID and secret to your code as a previous user who registered with the same instance (I have no idea if it does that or not) that would be an implementation detail and not something you should make any assumptions about. ([Here](https://github.com/mastodon/mastodon/issues/9707#issuecomment-1062282551) developers are advised to not add any random information per registration.)

If that is right (register per user per instance), I would also say this is a somewhat novel and unexpected use of OAuth, particularly for those of us who have experience with the more common single-registration scenario, and it really should be documented more explicitly in the API docs. (Eg: "Yes, this is different than other OAuth implementations you've encountered, but ...")

**Comment by MegaKeegMan** [comment]
This is a few years later, but I would also really appreciate more information on this. This does not seem sufficiently documented, but such documentation would be very valuable.
